### ISSUE TITLE
permutations strategy only works for small sequences

### ISSUE THREAD

**Issue by elvis-sik** [HypothesisWorks/hypothesis-python#438]
I am trying to test tictactoe-playing code. The player must have an ordered preference of positions to play. To try out several different combinations, I am passing a list of all positions to `st.permutations`. Then I expect it to spit out a list of the positions of the game in random order.
The problem I am having is that only small things work.

In this code, in my machine:

```python
SIZE = 50
@given(foo=st.permutations(range(SIZE)))
def test_hey(foo):
    assert 1
```

A failed health check will occur. If I change `SIZE` to `49`, it does not occur anymore.

I'd like to use it with much greater lists. I have no idea how feasible this would be. If this is not feasible, I'd be happy to know.

```
>>> import hypothesis as h
>>> h.__version__
'3.6.1'
```

**Comment by alexwlchan** [comment]
I can’t seem to reproduce this locally.

I’m using the following code, lifted from your example:

```python
# demo.py
from hypothesis import given
from hypothesis.strategies import permutations

SIZE = 50

@given(foo=permutations(range(SIZE)))
def test_hey(foo):
    assert True
```

and running `py.test demo.py`.

I wonder if it might be a Python 2/3 thing, because `range()` can be a generator-like object or a list, but I get a clean test pass on both versions.

Please could you share:

1. Which Python version you’re using
2. What OS you’re using
3. The full traceback (including the text of the healthcheck)

**Comment by elvis-sik** [comment]
Well, apparently today I have to tweak this to have `SIZE = 51` to reproduce. This is the case both for the file you provided and for the one I had (which had basically the same thing but inside a file with many tests).

I use Anaconda and Python 3:

```
$ python --version
Python 3.5.2 :: Anaconda custom (64-bit)
```

I use an up-to-date Arch Linux:

```
$ pacman -Q linux
linux 4.8.13-1
```

The traceback is this:

```
============================================== test session starts ==============================================
platform linux -- Python 3.5.2, pytest-3.0.5, py-1.4.31, pluggy-0.4.0
rootdir: /home/elvis/OpenSource/hypothesis-bug, inifile:
plugins: hypothesis-3.6.1
collected 1 items

demo.py F

=================================================== FAILURES ====================================================
___________________________________________________ test_hey ____________________________________________________

    @given(foo=permutations(range(SIZE)))
>   def test_hey(foo):

demo.py:8:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
../../miniconda3/envs/sci/lib/python3.5/site-packages/hypothesis/core.py:438: in wrapped_test
    HealthCheck.too_slow,
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

message = 'Data generation is extremely slow: Only produced 9 valid examples in 1.01 seconds (0 invalid ones and0 exceeded maxi... want to disable just this health check, add HealthCheck.too_slow to the suppress_health_checksettings for this test.'
label = <HealthCheck.too_slow: 3>

    def fail_health_check(message, label):
        if label in settings.suppress_health_check:
            return
        message += (
            '\nSee https://hypothesis.readthedocs.io/en/latest/health'
            'checks.html for more information about this. '
        )
        message += (
            'If you want to disable just this health check, add %s '
            'to the suppress_health_check settings for this test.'
        ) % (label,)
>       raise FailedHealthCheck(message)
E       hypothesis.errors.FailedHealthCheck: Data generation is extremely slow: Only produced 9 valid examples in 1.01 seconds (0 invalid ones and 0 exceeded maximum size). Try decreasing size of the data you're generating (with e.g.average_size or max_leaves parameters).
E       See https://hypothesis.readthedocs.io/en/latest/healthchecks.html for more information about this. If you want to disable just this health check, add HealthCheck.too_slow to the suppress_health_check settings for this test.

../../miniconda3/envs/sci/lib/python3.5/site-packages/hypothesis/core.py:306: FailedHealthCheck
=========================================== 1 failed in 1.34 seconds ============================================
```

`range` versus `list(range())` makes no difference, I think `permutations` does create a list out of the provideded sequence anyway: https://github.com/HypothesisWorks/hypothesis-python/blob/master/src/hypothesis/strategies.py#L860

**Comment by alexwlchan** [comment]
Okay, so I can reproduce this if I bump up `SIZE` to a truly ludicrous number (e.g. 1000), but not 51. Tipping point for me is 60.

I also see a slightly different error when I bump it up large:

> hypothesis.errors.FailedHealthCheck: **Examples routinely exceeded the max allowable size.** (9 examples overran while generating 0 valid ones). Generating examples this large will usually lead to bad results. You should try setting average_size or max_size parameters on your collections and turning max_leaves down on recursive() calls.

And if I turn it down so I’m close to the tipping point, I get other warnings about overly large examples.

I suspect the exact value where you overflow will depend a bit on your local setup – in particular, the CPU power that’s available when you run the test – but now I can reproduce it reliably.

Looking at [the source code](https://github.com/HypothesisWorks/hypothesis-python/blob/master/src/hypothesis/strategies.py#L857) for the `permutations` strategy, it indeed does cast to a list internally:

```python
def permutations(values):
    """Return a strategy which returns permutations of the collection
    "values"."""
    values = list(values)
    if not values:
        return just(()).map(lambda _: [])

    def build_permutation(swaps):
        initial = list(values)
        for i, j in swaps:
            initial[i], initial[j] = initial[j], initial[i]
        return initial
    n = len(values)
    index = integers(0, n - 1)
    return lists(tuples(index, index), max_size=n ** 2).map(build_permutation)
```

I wonder if we might be able to get by without the `initial = list(values)`, but thinking a bit more I suspect that might throw off the examples database when we try to save a failing example.

I don’t know what the “max allowable size” for a strategy means, or what causes Hypothesis to fall over here – perhaps @DRMacIver can enlighten us?

**Comment by alexwlchan** [comment]
Note: this comment:

> I expect it to spit out a list of the positions of the game in random order.

also makes me wonder whether there might be something in the [stateful testing area](http://hypothesis.readthedocs.io/en/latest/stateful.html) that might be a better fit than crunching something through the `permutations` strategy?

**Comment by DRMacIver** [comment]
"max allowable size" mostly only applies to collections types.

The problem you're probably running into here is that the permutations strategy requires a quadratically large amount of data to pick a good permutation in the current implementation. This is mostly to facilitate good shrinking (which a basic Fisher Yates style shuffle doesn't really do so well at).

I'm somewhat surprised that this is a problem at such low numbers though. I'll investigate whether a more linear approach is viable.

**Comment by DRMacIver** [comment]
It looks like the implementation of permutations came before a bunch of things that are now easy. So this was actually much simpler to fix than I expected.

**Comment by alexwlchan** [comment]
Resolved by #442.
